lib/git-lfs-s3/application.rb

```diff
@@ -47,6 +49,151 @@ def protected!
+        'size' => size,
+        'authenticated' => authenticated,
+        'actions' => {
+          'upload' => {
+            'href' => object.presigned_url(:put, acl: 'public-read', :expires_in => 86400)
+          }
+        },
+        'expires_at' => DateTime.now.next_day.to_time.utc.iso8601
+      }
+    end
+
+    def object_error(error, message, object, object_json)
+      # Format a single error object.
+      {
+        'oid' => object_json[:oid],
+        'size' => object_json[:size],
+        'error' => {
+          'code' => error,
+          'message' => message
+        }
+      }
+    end
+
+    def download(authenticated, params)
+      # Handle git-lfs batch downloads.
+      objects = Array.new
+      params[:objects].each do |object_json|
+        object_json = indifferent_params object_json
+        object = object_data object_json[:oid]
+        if valid_object? object_json
+          if object.exists?
+            objects.push object_download(authenticated, object, object_json)
+          else
+            objects.push object_error(404, 'Object does not exist', object, object_json)
+          end
+        else
+          objects.push object_error(422, 'Validation error', object, object_json)
+        end
+      end
+      objects
+    end
+
+    def upload(authenticated, params)
+      # Handle git-lfs batch uploads.
+      objects = Array.new
+      params[:objects].each do |object_json|
+        object_json = indifferent_params object_json
+        object = object_data object_json[:oid]
+        if valid_object? object_json
+          if object.exists?
+            objects.push object_download(authenticated, object, object_json)
+          else
+            objects.push object_upload(authenticated, object, object_json)
+          end
+        else
+          objects.push object_error(422, 'Validation error', object, object_json)
+        end
+      end
+      objects
+    end
+
+    def lfs_resp(objects)
+      # Successful git-lfs batch response.
+      status 200
+      resp = {
+        'transfer' => 'basic',
+        'objects' => objects
+      }
+      logger.debug resp
+      body MultiJson.dump(resp)
+    end
+    
+    def error_resp(status_code, message)
+      # Error git-lfs batch response.
+      status status_code
+      resp = {
+        'message' => message,
+        'request_id' => SecureRandom::uuid
+      }
+      logger.debug resp
+      body MultiJson.dump(resp)
+    end
+    
+    post '/objects/batch', provides: 'application/vnd.git-lfs+json' do
+      # git-lfs batch API
+      authenticated = authorized?
+      params = indifferent_params(JSON.parse(request.body.read))
+      
+      if params[:operation] == 'download'
+        objects = download(authenticated, params)
+      elsif params[:operation] == 'upload'
+        if authenticated
+          objects = upload(authenticated, params)
+        else
+          objects = nil
+        end
+      end
+      
+      if objects
```

Comment:
If `operation` isn't defined in the request, this will throw a `NameError` exception.  I recommend using a case statement that vomits on undefined `operations`.